src/Microsoft.OData.Client/ALinq/ResourceBinder.GroupByKeySelectorAnalyzer.cs

```diff
@@ -0,0 +1,118 @@
+            /// <param name="paramExpr">The parameter of the key selector expression, e.g., d1 => new { d1.ProductId, d1.CustomerId }.</param>
+            private GroupByKeySelectorAnalyzer(QueryableResourceExpression source, ParameterExpression paramExpr)
+            {
+                this.input = source;
+                this.lambdaParameter = paramExpr;
+            }
+
+            /// <summary>
+            /// Analyzes a GroupBy key selector for property or properties that the input sequence is grouped by. 
+            /// </summary>
+            /// <param name="input">The input resource expression.</param>
+            /// <param name="keySelector">Key selector expression to analyze.</param>
+            internal static void Analyze(QueryableResourceExpression input, LambdaExpression keySelector)
+            {
+                Debug.Assert(input != null, "input != null");
+                Debug.Assert(keySelector != null, "keySelector != null");
+
+                GroupByKeySelectorAnalyzer analyzer = new GroupByKeySelectorAnalyzer(input, keySelector.Parameters[0]);
+
+                MemberInitExpression memberInitExpr = keySelector.Body as MemberInitExpression;
+
+                if (memberInitExpr != null)
+                {
+                    analyzer.Visit(memberInitExpr);
+                }
+                else
+                {
+                    analyzer.Visit(keySelector.Body);
+                }
+            }
+
+            /// <inheritdoc/>
+            internal override NewExpression VisitNew(NewExpression nex)
+            {
+                // Maintain a mapping of grouping expression and respective member
+                // The mapping is cross-referenced if any of the grouping expression 
+                // is referenced in result selector
+                if (nex.Members != null && nex.Members.Count == nex.Arguments.Count)
+                {
+                    for (int i = 0; i < nex.Arguments.Count; i++)
+                    {
+                        this.input.Apply.KeySelectorMap.Add(nex.Members[i].Name, nex.Arguments[i]);
+                    }
+
+                }
+
+                return base.VisitNew(nex);
+            }
+
+            /// <inheritdoc/>
+            internal override MemberAssignment VisitMemberAssignment(MemberAssignment assignment)
+            {
+                // Maintain a mapping of grouping expression and respective member
+                // The mapping is cross-referenced if any of the grouping expression 
+                // is referenced in result selector
+                this.input.Apply.KeySelectorMap.Add(assignment.Member.Name, assignment.Expression);
+
+                return base.VisitMemberAssignment(assignment);
+            }
+
+            /// <inheritdoc/>
+            internal override Expression VisitMemberAccess(MemberExpression m)
+            {
+                MemberExpression memberExpr = StripTo<MemberExpression>(m);

```

Comment:
Why is this necessary if the argument is already a `MemberExpression`?

Reply:
what does StripTo do?

Reply:
@habbes Agreed and fixed. Strip had no value here
@ElizabethOkerio The `StripTo<T>` method strips the specified expression off intermediate expression (unnecessary converts and quotes) and returns the underlying expression of type T (or null if it's not of that type).